### PR TITLE
Version 0.3.6

### PR DESCRIPTION
**Pull request by abhinavsriva**
Canonical Step Time pipeline, integrations, and reliability hardening

## Summary

This release consolidates Step Time analysis behind one canonical pipeline, makes missing telemetry explicit, updates the versioned final-summary contract and compare behavior, expands integration coverage, and refreshes the public documentation and CI gates.

### Canonical Step Time analysis and presentation

- Replaced the split Step Time helpers with a typed, repository-backed  `traceml_ai.step_time` pipeline shared by live displays and final summaries.
- Preserved unavailable timing signals as `null` rather than treating them as   measured zeroes, and emits `INCOMPLETE_DATA` when the evidence is incomplete.
- Updated CLI and dashboard Step Time views for sparse phase coverage, cold startup, expiry of stale live data, rank-aware tables, and canonical clocks.
- Added executable contract, invariant, repository, pipeline, live-session,   and presentation tests for the new boundary.

### Final summaries and comparison

- Published the schema-1.7 canonical Step Time summary contract, including   selected-clock, traced-clock, CPU/GPU, and nullable timing fields.
- Updated HTML rendering and compare compatibility so historical reports keep their pre-1.7 interpretation while new reports use canonical Step Time.
- Added nullable-summary fixtures and tests to prevent `null` values from   becoming fabricated zeroes or leaking as raw `None` in HTML.

### Integrations and runtime behavior

- Added an Accelerate integration guide, minimal example, capability handling,   and real Hugging Face Trainer coverage for gradient accumulation and H2D   timing windows.
- Added telemetry conformance coverage across supported integration paths.
- Corrected the public `traceml.start()` alias so it forwards the same  missing-aggregator policy as `traceml.init()`.
- Added a real in-process TCP client-to-server-to-aggregator-to-SQLite   round-trip suite, covering valid frames, corrupt-frame recovery, unknown   sampler routing, and lazy reconnect after server restart.
- Hardened atomic file-write cleanup: exceptions during write, flush, fsync,   or replace preserve the previous target and remove the sibling temp file.

### Documentation, CI, and test hygiene

- Refreshed the README, quickstart, API reference, architecture,   troubleshooting, distributed-training, and Step Time contract documentation.
- Added strict docs validation for documentation-only pull requests and made   the integration job install the Hugging Face extra so its tests run rather   than skip.
- Removed a module-import Lightning exploration that ran real training,   printed output, and asserted no TraceML behavior.
- Kept sparse Step Time renderer coverage consolidated in its existing suite;   no duplicate test file remains.

## Compatibility notes

- `final_summary.json` schema 1.7 uses canonical `step_time_ms` and nullable   metric fields. Readers retain compatibility handling for reports produced   before 1.7.
- The TCP client does not run a background reconnect loop. After a failed send,   its next send lazily attempts a new connection.

## Validation

```text
PYTHONDONTWRITEBYTECODE=1 pytest -p no:cacheprovider -q \
  tests/config tests/core tests/step_time tests/diagnostics \
  tests/reporting/test_view.py tests/telemetry
```

Result: `395 passed`.

```text
PYTHONDONTWRITEBYTECODE=1 pytest -p no:cacheprovider -q \
  tests/aggregator tests/integrations tests/runtime tests/sdk \
  tests/reporting/compare tests/reporting/html tests/reporting/summary \
  tests/renderers tests/display tests/test_h2d_timing.py
```

Result: `519 passed, 2 skipped`.

`git diff --check main...HEAD` passes.

## Summary by CodeRabbit

* **New Features**
  * Unified Step Time analysis across CLI, dashboard, and summaries.
  * Added clearer CPU/GPU timing metrics, clock-aware comparisons, and live-data freshness handling.
  * Added Hugging Face Accelerate integration guidance and examples.
  * Public instrumentation helpers now have typed, documented interfaces.

* **Bug Fixes**
  * Missing measurements are reported as unavailable or incomplete instead of zero.
  * Improved live-session reuse and dashboard handling of expired data.

* **Documentation**
  * Summary mode is now the default; live modes are explicitly selectable.
  * Updated schema 1.7, comparison, quickstart, troubleshooting, and Step Time documentation.

### CODE REVIEW

**Comment by Pendu** [comment]
## Pre-merge validation of version_0.3.6 (two passes: 8162aff full CPU+GPU, c2bd8b6 CPU after #309/#311/#313)

Since this branch batches a large step-time refactor plus the missing-signals
stack, I ran a structured validation pass before we merge to main. Summary:
**everything measured is green**; follow-up issues filed (#315,
#316, plus test-debt #317 #318 #319 #320); one ack needed
from @abhinavsriva below.

### CPU gate (both at 8162aff and re-run at c2bd8b6 after the naming PRs)

- Full suite with the `hf` extra, exact CI invocations: core 388 passed,
  integration 490 passed + 2 known skips (at c2bd8b6).
- All 7 CPU-runnable examples through the real launcher, default flags:
  exit 0, `schema_version` 1.7, run manifest `completed`, no `[TraceML]`
  errors. The bottleneck demo flips its verdict correctly
  (fast: BALANCED, slow: INPUT-BOUND with measured input wait), Lightning and
  HF both emit measured `input_wait`, and
  `ddp_minimal.py --nproc-per-node=2` (gloo on CPU) reports 2/2 ranks with
  per-rank rows.
- A/B against current main on quickstart + the HF example: the step_time
  sections have identical key sets and values within run-to-run noise; the
  only semantic difference is `h2d_ms` `0.0` -> `null`, which is the
  documented 1.6 -> 1.7 nullable change. `traceml compare` across the two
  versions renders `H2D 0.0 ms | n/a | n/a` with no fabricated delta.
- Old-artifact compatibility: `view`, HTML render, and `compare` against a
  schema-1.4 summary from June all work, with the cross-version warning and
  an honest INCONCLUSIVE verdict.
- Mid-run `traceml.summary()` round-trip works (card + compact dict), and the
  run continues to a correct end-of-run summary.
- The new live tail query uses `idx_step_time_samples_global_rank_step_id`
  (checked with EXPLAIN QUERY PLAN on a launcher-produced database), and the
  pipeline benchmark shows a 32-rank analyze at ~110-128 ms with cache hits
  ~12 ms, well inside the 2 s display tick.

### GPU pass (4x T4, at 8162aff)

- DDP straggler demo, 2 GPUs, all three scenarios: balanced run shows no
  straggler verdict; induced input straggler on rank 1 is diagnosed
  INPUT_STRAGGLER with culprit r1, victim r0, ~76% impact and the average
  input wait matching the injected sleep exactly; induced compute straggler
  on rank 1 is diagnosed STRAGGLER with culprit r1.
- FSDP smoke on 2 GPUs completes with 2/2 ranks and a sane summary.
- `h2d` is present on the DDP/FSDP runs (gpu clock) and `null` on the HF
  Trainer path, consistent with #276; an independent CUDA-event timing of the
  real transfer (29.3 ms) confirms the `n/a` is honest rather than a missed
  measurement.
- Dashboard visual pass on a live run: hero ribbon, engine-sourced
  diagnostics rail, step-memory section, and the sparse-signals state
  (hatched slivers + "partial: FWD,RESIDUAL" label + residual share n/a) all
  render; teardown shows a clean "Connection lost" toast. CLI live mode
  populates the IW/H2D/FWD/BWD/OPT/STEP table and transitions honestly from
  NO DATA through BALANCED to INPUT-BOUND as the window fills.

### Protocol lines

GATE: CPU gate PASS at 8162aff and re-run PASS at c2bd8b6 | GPU pass PASS at 8162aff (not re-run after the naming PRs, per our chat; manual spot-check before merge is yours)
AB: quickstart + HF vs main, key-identical, only documented h2d null change
RANKS: 2-GPU DDP straggler culprit correctness verified (input + compute), FSDP 2/2
VISUAL: dashboard + CLI passes done at 8162aff; sparse-signals state verified

### Needs your ack

1. The single-node default mode change dashboard -> summary (from #292):
   intentional as shipped? It also means the live surfaces get less default
   exposure after merge, which raises the value of #315.
2. #316 (quickstart has no DataLoader, so first-run reports
   INCOMPLETE_DATA): fine to fix post-merge, but worth deciding before the
   next release.

Findings that are NOT blockers: #315 (dashboard sections unreachable at
~1280x720), and a single unreproduced "Message too long" websocket toast at
one dashboard teardown (watching, no issue filed).
